output/cloud/expv2/hdr.go

```diff
@@ -0,0 +1,270 @@
+	// with a significant counter in the Buckets slice (a not zero value).
+	// In this way, all the buckets after can be omitted.
+	LastNotZeroBucket uint32
+
+	// Max is the absolute maximum observed value.
+	Max float64
+
+	// Min is the absolute minimum observed value.
+	Min float64
+
+	// Sum is the sum of all observed values.
+	Sum float64
+
+	// Count is counts the amount of observed values.
+	Count uint32
+}
+
+// newHistogram creates an histogram of the provided values.
+//
+// TODO: after the aggregation layer probably this constructor
+// doesn't make any sense in this way.
+// It aggregates calling addToBucket time-to-time so
+// trimzeros has to be called at the end of the process.
+func newHistogram(values []float64) histogram {
+	h := histogram{}
+	if len(values) < 1 {
+		return h
+	}
+
+	for i := 0; i < len(values); i++ {
+		h.addToBucket(values[i])
+	}
+
+	h.trimzeros()
+	return h
+}
+
+// addToBucket increments the counter of the bucket
+// releated to the provided value.
+// If the value is lower or higher than the trackable limits
+// then it is counted into specific buckets.
+// All the stats are also updated accordingly.
+//
+// TODO: add the test case in units doing
+// addToBucket + trimzeros + addToBucket
+// so calling addToBucket after trimzeros was called.
+// We don't expect to have this case but the current API
+// would support it so we need to make sure it works or refactor.
+func (h *histogram) addToBucket(v float64) {
+	if h.Count == 0 {
+		h.Max, h.Min = v, v
+	} else {
+		if v > h.Max {
+			h.Max = v
+		}
+		if v < h.Min {
+			h.Min = v
+		}
+	}
+
+	h.Count++
+	h.Sum += v
+
+	if v > highestTrackable {
+		h.ExtraHighBucket++
+		return
+	}
+	if v < lowestTrackable {
+		h.ExtraLowBucket++
+		return
+	}
+
+	index := resolveBucketIndex(v)
+	blen := uint32(len(h.Buckets))
+	if blen == 0 {
+		h.FirstNotZeroBucket = index
+		h.LastNotZeroBucket = index
+	} else {
+		if index < h.FirstNotZeroBucket {
+			h.FirstNotZeroBucket = index
+		}
+		if index > h.LastNotZeroBucket {
+			h.LastNotZeroBucket = index
+		}
+	}
+
+	if index >= blen {
+		h.grow(index)
+	}
+	h.Buckets[index]++
+}
+
+// grow expands the buckets slice
+// with zeros up to the required index
+func (h *histogram) grow(index uint32) {
+	i := int(index)
+	if len(h.Buckets)-1 > i {
+		panic("buckets is already bigger than requested index")
+	}
+	if cap(h.Buckets) > i {
+		// See https://go.dev/ref/spec#Slice_expressions
+		// "For slices, the upper index bound is
+		// the slice capacity cap(a) rather than the length"
+		h.Buckets = h.Buckets[:index+1]
+	} else {
+		length := i + 1
+		// let's make two times larger of
+		// the current request
+		newBuckets := make([]uint32, length, (len(h.Buckets)+length)*2)
+		copy(newBuckets, h.Buckets)
+		h.Buckets = newBuckets
+	}
+}
+
+// trimzeros removes all buckets that have a zero value
+// from the begin and from the end until
+// the first not zero bucket.
+func (h *histogram) trimzeros() {
```

Comment:
I've noticed in other PRs using this that this method had to be explicitly called here and there.

Just sharing a thought I had and I don't have an alternative proposal to offer, yet it felt like the potential for future bugs "we had this bug because we didn't call `trimzeroes` as we should have", and I wondered if there might be opportunities to avoid having to call it explicitly.

Is that relevant?